src/main/java/com/yahoo/bullet/bql/extractor/ProjectionExtractor.java

```diff
@@ -5,61 +5,175 @@
+        }
+        // Perform a copy when there are additional fields selected
+        return new Projection();
+    }
+
+    /**
+     * Project the select fields without their aliases since they'll be renamed in the aggregation instead.
+     */
+    private Projection extractDistinct() {
+        List<ExpressionNode> expressions = processedQuery.getSelectNodes().stream().map(SelectItemNode::getExpression).collect(Collectors.toList());
+        // If all select fields are simple field expressions, then there's no need to project anything.
+        // Order by computations are done afterward and won't clobber the original record since the aggregation creates a new record.
+        if (expressions.stream().allMatch(processedQuery::isSimpleFieldExpression)) {
+            return null;
+        }
+        List<Projection.Field> fields = expressions.stream().map(this::toNonAliasedField).collect(Collectors.toList());
+        return new Projection(fields);
+    }
 
     /**
-     * Constructor that requires a set of selectFields and map of aliases.
-     *
-     * @param selectFields The non-null Set of selected fields.
-     * @param aliases      The non-null Map of aliases.
-     * @throws NullPointerException when any of selectFields and aliases is null.
+     * Project group by fields and any fields with aggregates (without their aliases since they'll be renamed in
+     * the aggregation). We care about select or order by fields, since those are computed afterwards.
      */
-    public ProjectionExtractor(Set<Expression> selectFields, Map<Node, Identifier> aliases) throws NullPointerException {
-        requireNonNull(selectFields);
-        requireNonNull(aliases);
+    private Projection extractGroup() {
+        Set<ExpressionNode> requiredNodes = new HashSet<>(processedQuery.getGroupByNodes());
+
+        // Project aggregate fields (need to filter for non-null because COUNT(*) does not have a field)
+        requiredNodes.addAll(processedQuery.getGroupOpNodes().stream().map(GroupOperationNode::getExpression)
+                                                                      .filter(Objects::nonNull)
+                                                                      .collect(Collectors.toSet()));
 
-        this.selectFields = selectFields;
-        this.aliases = aliases;
+        // If the group by and aggregate fields are all simple field expressions, then we don't need to project anything
+        if (requiredNodes.stream().allMatch(processedQuery::isSimpleFieldExpression)) {
+            return null;
+        }
+        List<Projection.Field> fields = requiredNodes.stream().map(this::toNonAliasedField).collect(Collectors.toList());
+        return new Projection(fields);
     }
 
     /**
-     * Extract a Projection.
-     *
-     * @return A Projection based on selectFields and aliases.
+     * Project count distinct fields. Ignore any possible aliasing, e.g. SELECT abc AS def, COUNT(DISTINCT abc) FROM STREAM();
```

Comment:
Do you want to do it that way? Computations don't really have meaning on a single count field so it can error out on additional fields. It can also error out on HAVING. (It already errors out on ORDER BY and LIMIT)

Reply:
Well it can go in that intermediate visitor that error checks.